idaes/generic_models/unit_models/heat_exchanger_lc.py

```diff
@@ -0,0 +1,203 @@
+        def ua_total_eq(b, t):
+            return b.overall_heat_transfer_coefficient[t] * b.area == 1 / (
+                    1 / b.ua_hot_side[t] + b.thermal_fouling_hot_side +
+                    b.thermal_resistance_wall + b.thermal_fouling_cold_side +
+                    1 / b.ua_cold_side[t])
+
+        @self.Constraint(
+            self.flowsheet().config.time,
+            doc='Wall temperature equation'
+        )
+        def wall_temperature_eq(b, t):
+            return b.temperature_wall[t] == 0.5 * (
+                   b.hot_side.properties_in[t].temperature +
+                   b.hot_side.properties_out[t].temperature) \
+                   + b.hot_side.heat[t] / b.ua_hot_side_to_wall[t]
+
+    def activate_dynamic_heat_eq(self):
+        """
+        Adds a heat holdup term to the overall energy balance for transient
+        simulations. Should only be used with dynamic flowsheets.
+
+        Args:
+            None
+        Returns:
+            None
+        """
+        self.unit_heat_balance.deactivate()
+        self.dynamic_heat_balance.activate()
+
+    def deactivate_dynamic_heat_eq(self):
+        """
+        Removes the heat holdup term from the overall energy balance,
+        restoring the behavior of the base 0D HeatExchanger model. This may
+        be useful for more complex models which need to be solved in several
+        stages.
+
+        Args:
+            None
+        Returns:
+            None
+        """
+        self.dynamic_heat_balance.deactivate()
+        self.unit_heat_balance.activate()
+
+    def build(self):
+        """
+        Building model
+
+        Args:
+            None
+        Returns:
+            None
+        """
+
+        super().build()
+        self._add_wall_variables()
+        self._add_wall_variable_constraints()
+
+        if self.flowsheet().config.dynamic:
+
+            s1_metadata = self.config.hot_side_config.property_package. \
+                get_metadata()
+            temp_units = s1_metadata.get_derived_units("temperature")
+            time_units = s1_metadata.get_derived_units("time")
+
+            self.dT_wall_dt = DerivativeVar(
+                self.temperature_wall,
+                wrt=self.flowsheet().config.time,
+                doc='Derivative of wall temperature with respect to time',
+                units=temp_units / time_units
+            )
+            @self.Constraint(
+                self.flowsheet().time,
+                doc='Unit heat balance equation with the added capacitance term'
+            )
+            def dynamic_heat_balance(b, t):
+                return b.hot_side.heat[t] + b.cold_side.heat[t] == \
+                       -b.dT_wall_dt[t] * b.heat_capacity_wall
+
+            self.activate_dynamic_heat_eq()
+            self.dT_wall_dt[self.flowsheet().config.time.first()].fix(0)
+            self.dT_wall_dt[:].value = 0
+
+    def initialize(self, *args, **kwargs):
+
+        if self.flowsheet().config.dynamic:

```

Comment:
The reason for this is that I want to use control volumes with `dynamic=False`. And the way the base class is written, the control volumes inherit their `dynamic` flags from the unit model config. In `heat_exchanger.py` we have:

```
        hot_side = _make_heater_control_volume(
            self,
            config.hot_side_name,
            config.hot_side_config,
            dynamic=config.dynamic,
            has_holdup=config.has_holdup,
        )
```

What you're suggesting would call for a config like so:

```
        m.fs.unit = HeatExchangerLumpedCapacitance(default={
            "shell": {"property_package": m.fs.properties,
                      "dynamic": False},
            "tube": {"property_package": m.fs.properties,
                     "dynamic": False},
            "dynamic": True})
```

And then, for instance, in `heat_exchanger.py`:

```
        try:
            hot_dynamic = hot_side_config.dynamic
        except KeyError:
            hot_dynamic = config.dynamic
        
        hot_side = _make_heater_control_volume(
            self,
            config.hot_side_name,
            config.hot_side_config,
            dynamic=hot_dynamic,
            has_holdup=config.has_holdup,
        )
```

That would give us three different levels where the `dynamic` flag could be configured: flowsheet -> unit model -> control volume. And in that case, we should probably do the same for `has_holdup`. Actually, the new logic probably belongs in `_process_config()`.

I'd be happy to try this, but I thought it best not to make any changes to `heat_exchanger.py` in case someone else is also working on it. Let me know which you prefer, to leave it as-is, or to make the changes to both unit models and have `dynamic` be configurable at each level.

Reply:
First, I think we need to work out exactly what you want to allow for, as I am not quite sure what is happening. Is it that you want to allow separate the decision about holdup dynamics from that of the wall energy balance then - i.e. allow the wall energy balance to be dynamic even if the holdups are steady-state?

Reply:
> Is it that you want to allow separate the decision about holdup dynamics from that of the wall energy balance then - i.e. allow the wall energy balance to be dynamic even if the holdups are steady-state?

Yes exactly!

Reply:
OK, my first thought then is that we should allow the user to choose wall dynamics separately from the material balance -  I.e. have a separate configuration option for the wall dynamics that the user can set, and then have some logic on the default setting (much like how the material dynamics are done).

We might want to think about any combinations that do not make sense however - I am not sure that there are any however (at most it would be setting everything to steady-state, in which case this devolves into the basic heat exchanger, and that will still work).

Reply:
> OK, my first thought then is that we should allow the user to choose wall dynamics separately from the material balance - I.e. have a separate configuration option for the wall dynamics that the user can set, and then have some logic on the default setting (much like how the material dynamics are done).

This is a nice solution. If there are no other concerns or objections, I'll go ahead and implement it. Do you have a preference for the new variable name? I was thinking either `dynamic_wall` or `dynamic_heat_balance`. Also, are there any other similar unit models? Because it might be good to standardize or reuse this variable name if possible.

> We might want to think about any combinations that do not make sense however - I am not sure that there are any however (at most it would be setting everything to steady-state, in which case this devolves into the basic heat exchanger, and that will still work).

Agreed. That did occur to me. My initial thought was that there wasn't a use case for having everything be steady-state, since one could just use the base heat exchanger model. But with this model, you at least get the wall temperature in addition. So that might still be useful to have. Especially if you're troubleshooting a flowsheet or model that's difficult to solve, it could be helpful to see this temperature at steady-state.

Also, if you're in "troubleshooting mode" I think it's better to change something with a flag rather than swapping out the entire unit model. Since using a different unit model introduces more unknowns. So I think it's important to make sure this model still works with steady-state flowsheets.

The same argument could be made for the other config combinations. The user might want to make sure their model solves with dynamic control volumes before activating the wall dynamics. And vice versa too.

Reply:
`dynamic_heat_balance` is the best of anything I can think of so far.

Reply:
> `dynamic_heat_balance` is the best of anything I can think of so far.

Ok done! I also added a test to make sure each combination of dynamic / dynamic_heat_balance can be built. And some xfail-ing tests for invalid configurations.